File: ports/csharp/src/gui/widgets/BcsWidget.cs
using System;
using Bcs.Core.Kernel;

namespace Bcs.Gui.Widgets
{
    public class BcsWidget : BcsObject
    {
        private bool _visible = false;
        private bool _enabled = true;
        private int _width = 0;
        private int _height = 0;
        private int _x = 0;
        private int _y = 0;
        private bool _hasFocus = false;
        private readonly object _lock = new object();

        public BcsWidget(BcsWidget parent = null) : base(parent)
        {
        }

        public void Show() { lock (_lock) { _visible = true; } }
        public void Hide() { lock (_lock) { _visible = false; } }
        public bool IsVisible { get { lock (_lock) { return _visible; } } }

        public void Resize(int width, int height)
        {
            lock (_lock)
            {
                _width = width;
                _height = height;
            }
        }

        public void Move(int x, int y)
        {
            lock (_lock)
            {
                _x = x;
                _y = y;
            }
        }

        public bool IsEnabled
        {
            get { lock (_lock) { return _enabled; } }
            set { lock (_lock) { _enabled = value; } }
        }

        public bool HasFocus
        {
            get { lock (_lock) { return _hasFocus; } }
            set { lock (_lock) { _hasFocus = value; } }
        }

        public override bool Event(BcsEvent e)
        {
            switch (e.Type)
            {
                case EventType.MouseButtonPress:
                    MousePressEvent(e);
                    return true;
                case EventType.MouseButtonRelease:
                    MouseReleaseEvent(e);
                    return true;
                case EventType.KeyPress:
                    KeyPressEvent(e);
                    return true;
                case EventType.KeyRelease:
                    KeyReleaseEvent(e);
                    return true;
                case EventType.Paint:
                    PaintEvent(e);
                    return true;
                default:
                    return base.Event(e);
            }
        }

        protected virtual void MousePressEvent(BcsEvent e) {}
        protected virtual void MouseReleaseEvent(BcsEvent e) {}
        protected virtual void KeyPressEvent(BcsEvent e) {}
        protected virtual void KeyReleaseEvent(BcsEvent e) {}
        protected virtual void PaintEvent(BcsEvent e) {}
    }
}

File: ports/rust/src/gui/widgets/bcs_widget.rs
use std::sync::{Arc, Mutex};
use crate::core::kernel::bcs_object::BcsObject;
use crate::core::kernel::bcs_event::{BcsEventT, EventType};

pub trait BcsWidgetImpl: Send + Sync {
    fn mouse_press_event(&self, _event: &dyn BcsEventT) {}
    fn mouse_release_event(&self, _event: &dyn BcsEventT) {}
    fn key_press_event(&self, _event: &dyn BcsEventT) {}
    fn key_release_event(&self, _event: &dyn BcsEventT) {}
    fn paint_event(&self, _event: &dyn BcsEventT) {}
}

pub struct BcsWidgetBaseImpl;
impl BcsWidgetImpl for BcsWidgetBaseImpl {}

pub struct BcsWidget {
    pub base: Arc<BcsObject>,
    visible: Mutex<bool>,
    enabled: Mutex<bool>,
    width: Mutex<i32>,
    height: Mutex<i32>,
    x: Mutex<i32>,
    y: Mutex<i32>,
    has_focus: Mutex<bool>,
    impl_delegate: Mutex<Box<dyn BcsWidgetImpl>>,
}

impl BcsWidget {
    pub fn new(parent: Option<Arc<BcsObject>>) -> Arc<Self> {
        let base = if let Some(p) = parent {
            BcsObject::new_with_parent(p)
        } else {
            BcsObject::new()
        };

        Arc::new(Self {
            base,
            visible: Mutex::new(false),
            enabled: Mutex::new(true),
            width: Mutex::new(0),
            height: Mutex::new(0),
            x: Mutex::new(0),
            y: Mutex::new(0),
            has_focus: Mutex::new(false),
            impl_delegate: Mutex::new(Box::new(BcsWidgetBaseImpl)),
        })
    }

    pub fn set_impl(&self, delegate: Box<dyn BcsWidgetImpl>) {
        *self.impl_delegate.lock().unwrap() = delegate;
    }

    pub fn show(&self) { *self.visible.lock().unwrap() = true; }
    pub fn hide(&self) { *self.visible.lock().unwrap() = false; }
    pub fn is_visible(&self) -> bool { *self.visible.lock().unwrap() }

    pub fn resize(&self, width: i32, height: i32) {
        *self.width.lock().unwrap() = width;
        *self.height.lock().unwrap() = height;
    }

    pub fn move_to(&self, x: i32, y: i32) {
        *self.x.lock().unwrap() = x;
        *self.y.lock().unwrap() = y;
    }

    pub fn set_enabled(&self, enabled: bool) { *self.enabled.lock().unwrap() = enabled; }
    pub fn is_enabled(&self) -> bool { *self.enabled.lock().unwrap() }

    pub fn set_focus(&self, focus: bool) { *self.has_focus.lock().unwrap() = focus; }
    pub fn has_focus(&self) -> bool { *self.has_focus.lock().unwrap() }

    pub fn event(&self, event: &dyn BcsEventT) -> bool {
        let delegate = self.impl_delegate.lock().unwrap();
        match event.event_type() {
            EventType::MouseButtonPress => {
                delegate.mouse_press_event(event);
                true
            },
            EventType::MouseButtonRelease => {
                delegate.mouse_release_event(event);
                true
            },
            EventType::KeyPress => {
                delegate.key_press_event(event);
                true
            },
            EventType::KeyRelease => {
                delegate.key_release_event(event);
                true
            },
            EventType::Paint => {
                delegate.paint_event(event);
                true
            },
            _ => self.base.event(event),
        }
    }
}
